app/image/gc/gc.go

```diff
@@ -67,111 +123,331 @@ func (g *imgGC) spin() {
+	if dryRun {
+		return
+	}
+
+	err = sweepOldImages()
+	if err != nil {
+		err = errors.Wrap(err, "errors running GC sweep")
 	}
+
+	return
 }
 
-func cleanImageForApp(a *app.App, imgName string, removeFromRegistry bool) bool {
-	shouldRemove := true
-	defer func() {
-		log.Debugf("[image gc] image %q processed, removed from registry: %v, removed from database: %v", imgName, removeFromRegistry, removeFromRegistry && shouldRemove)
-	}()
-	err := a.CleanImage(imgName)
+func CleanImage(appName string, version appTypes.AppVersionInfo, removeFromRegistry bool) {
+	a, err := app.GetByName(appName)
 	if err != nil {
-		shouldRemove = false
-		log.Errorf("[image gc] error removing old image from provisioner for app %q: %v. Image kept on list to retry later.",
-			imgName, err.Error())
+		log.Errorf("[image gc] error getting app by name %q: %v. Image kept on list to retry later.", appName, err)
+		return
 	}
-	if removeFromRegistry {
-		if err := registry.RemoveImageIgnoreNotFound(imgName); err != nil {
-			log.Errorf("[image gc] error removing old image from registry %q: %s. Image kept on list to retry later.",
-				imgName, err.Error())
-			return false
-		}
-		return shouldRemove
+	pruned := pruneVersionFromProvisioner(a, version)
+	if !pruned || !removeFromRegistry {
+		return
 	}
-	return false
+
+	pruned = pruneVersionFromRegistry(version)
+	if !pruned {
+		return
+	}
+
+	pruneVersionFromStorage(appName, version)
 }
 
-func removeOldImages() error {
-	log.Debugf("[image gc] starting image gc process")
-	defer log.Debugf("[image gc] finished image gc process")
+func markOldImages() error {
+	gcExecutionsTotal.WithLabelValues("mark").Inc()
+	timer := prometheus.NewTimer(executionDuration.WithLabelValues("mark"))
+	defer timer.ObserveDuration()
+
+	log.Debugf("[image gc] starting gc process to select old images")
+	defer log.Debugf("[image gc] finished gc process to select old images")
 	allAppVersions, err := servicemanager.AppVersion.AllAppVersions()
 	if err != nil {
 		return err
 	}
 	historySize := image.ImageHistorySize()
 	multi := tsuruErrors.NewMultiError()
-	for _, versions := range allAppVersions {
-		log.Debugf("[image gc] processing %d versions for app %q", len(versions.Versions), versions.AppName)
-		a, err := app.GetByName(versions.AppName)
+	for _, appVersions := range allAppVersions {
+		log.Debugf("[image gc] processing %d versions for app %q", len(appVersions.Versions), appVersions.AppName)
+		a, err := app.GetByName(appVersions.AppName)
 		if err != nil && err != appTypes.ErrAppNotFound {
 			multi.Add(err)
 			continue
 		}
 		if a == nil {
-			log.Debugf("[image gc] app %q not found, removing everything", versions.AppName)
-			err = registry.RemoveAppImages(versions.AppName)
-			if err != nil {
-				multi.Add(err)
-			}
-			err = servicemanager.AppVersion.DeleteVersions(versions.AppName)
+			log.Debugf("[image gc] app %q not found, removing everything", appVersions.AppName)
+			err = pruneAllVersionsByApp(appVersions.AppName)
```

Comment:
Since we're separating mark and sweep this seems a bit out of place here. Wouldn't it be better to only mark all these versions instead of sweeping straight away?